x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/Reverse.java

```diff
@@ -0,0 +1,124 @@
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.Example;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlConfigurationFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.session.Configuration;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.DEFAULT;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isString;
+
+public class Reverse extends EsqlConfigurationFunction {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Reverse", Reverse::new);
+
+    private final Expression field;
+
+    @FunctionInfo(
+        returnType = { "keyword", "text" },
+        description = "Returns a new string representing the input string in reverse order.",
+        examples = {
+            @Example(file = "string", tag = "reverse"),
+            @Example(file = "string", tag = "reverseEmoji", description = "`REVERSE` works with unicode, too!") }
+    )
+    public Reverse(
+        Source source,
+        @Param(
+            name = "str",
+            type = { "keyword", "text" },
+            description = "String expression. If `null`, the function returns `null`."
+        ) Expression field,
+        Configuration configuration
+    ) {
+        super(source, List.of(field), configuration);
+        this.field = field;
+    }
+
+    private Reverse(StreamInput in) throws IOException {
+        this(Source.EMPTY, in.readNamedWriteable(Expression.class), ((PlanStreamInput) in).configuration());
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(field());
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return field.dataType();
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        return isString(field, sourceText(), DEFAULT);
+    }
+
+    @Override
+    public boolean foldable() {
+        return field.foldable();
+    }
+
+    @Evaluator
+    static BytesRef process(BytesRef val, @Fixed Locale locale) {
+        return BytesRefs.toBytesRef(new StringBuilder(val.utf8ToString()).reverse());
```

Comment:
I'm quite curious about `StringBuilder`'s behavior when it reverses combining marks. I'm reasonably sure it'll do the right thing with high surrogates because it's in the jvm - and I'm sure some your tests see that one. But it'd be nice to see what happens for combining marks. We don't have to figure out the right thing - but we should know what it does and document it.

Reply:
Hi @nik9000 ... I don't know what a mark is :D

Reply:
https://en.wikipedia.org/wiki/Combining_Diacritical_Marks

Reply:
`row text = "XxX̄x̄ abc 123 XxX̄x̄ " | eval x = reverse(text), y = reverse(x), z = y == text`

Apart from screwing up our text representation, maybe it doesn't work so well?

```
         text         |          x           |          y           |       z       
----------------------+----------------------+----------------------+---------------
XxX̄x̄ abc 123 XxX̄x̄ | ̄x̄XxX 321 cba ̄x̄XxX|XxX̄x̄ abc 123 XxX̄x̄ |true
```

Reply:
Hmm, so the double operation gets us back to the right state, but a single reverse leaves the bar floating behind the `x`. Did I understand this correctly?

Reply:
Actually I think this behavior may be correct at least in the most technical sense...

`XxX̄x̄ ` is `\u0058\u0078\u0058\u0304\u0078\u0304\u0020` where `\u0304` is the combining mark.

Reversing these, mark by mark, you get

`\u0020\u0304\u0078\u0304\u0058\u0078\u0058` which leaves the (now) first combining mark in front of the space (`\u0020`) instead of the little `x` (`\u0078`. This results in ` ̄x̄XxX`

Reply:
Yeah, I think your understanding is correct. From user point-of-view I believe this behavior is incorrect. @nik9000 might know more about this, but I would explore other ways of reversing a string to have it do the right thing the user expects.

Reply:
I don't know what the right behavior when reversing a string containing combining marks is. I think it'd take some research and blow out the time this PR takes by a lot. It looks like postgres's REVERSE blows up on utf-8 text in general - or - it does in dbfiddle. MySQL seems to do ok.

So! I think we should say in the docs "if the string contains [combining marks](https://en.wikipedia.org/wiki/Combining_Diacritical_Marks) the behavior is not defined." We should have a test that makes sure it does *something* and doesn't crash or make invalid utf-8. Someone's got to confirm that it's still valid utf-8 if the combining marks are reversed, but otherwise I'm ok merging with whatever behavior for combining marks. It's very important the REVERSE makes valid utf-8. I don't know the best way to check that, but `assert result.utf8ToString() : result + " is valid utf8";` should do it - though maybe I've written the `assert` backwards - I can never remember the order.

I think I mostly wanted us to know what it does with these "fun" bits of unicode. We know it properly works for non-ascii, and bytes that need high surrogates in utf-16. So long as we don't actively crash for combining marks I'm ok for now.

Reply:
Ok. I gave this problem to the AI overlords and this is what I got

<details>
```java
import java.text.BreakIterator;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class UnicodeStringReverser {

    public static String reverse(String input) {
        BreakIterator boundary = BreakIterator.getCharacterInstance(Locale.getDefault());
        boundary.setText(input);

        List<String> characters = new ArrayList<>();
        int start = boundary.first();
        for (int end = boundary.next(); end != BreakIterator.DONE; start = end, end = boundary.next()) {
            characters.add(input.substring(start, end));
        }

        StringBuilder reversed = new StringBuilder(input.length());
        for (int i = characters.size() - 1; i >= 0; i--) {
            reversed.append(characters.get(i));
        }

        return reversed.toString();
    }

    public static void main(String[] args) {
        String input = "\u0058\u0078\u0058\u0304\u0078\u0304\u0020";
        String reversed = reverse(input);
        System.out.println(reversed);
    }
}
```
</details>

But, if we're not sure we'd want to keep the marks ahead of the others, I can just add the disclaimer.

Reply:
Would keeping the combining marks in the original order ensure utf8 validity?

Reply:
Tangentially, this discussion reminds me of a time when I sent "🤦‍♂️" in an email through the MS Outlook client to a group of corporate high-ups at another company. The MS Outlook client expanded it to something like 🤦 ♂️. That was when I _actually_ face-palmed

Reply:
So utf-8 validity and unicode validity are two different things. utf-8 is valid so long as the combining marks don't get sliced apart. It looks like they don't so we're good there.

I think a `NOTE` about combining marks being broken from the original combination character is probably good here. I'm not an expert in this stuff, but it's ok on the rest of ESQL itself if you reverse the combining marks. In other words - the behavior that you have now is fine - let's just document the combining marks are weird. At some point we'll have to pull in unicode normalization functions, but that's a problem for later.

Reply:
Yeah, so the simple byte-wise reverse messes up grapheme clusters which includes a bunch of emojis, but also some natural language characters such as क्‍ष. 

@nik9000 and I synced and decided that it's worth handling this edge case, along with the combining marks.

To limit the performance impact, I will scan the string for non-ascii bytes and if I don't find any, we'll go with the simple reverse. Otherwise, we'll accept the hit that will come from the `BreakIterator`. 

I will update the PR.

Reply:
Done now!